src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/collectors/NodeStatsAllShardsMetricsCollector.java

```diff
@@ -0,0 +1,257 @@
+ * A copy of the License is located at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package com.amazon.opendistro.elasticsearch.performanceanalyzer.collectors;
+
+import java.lang.reflect.Field;
+import java.util.HashMap;
+import java.util.Map;
+import com.amazon.opendistro.elasticsearch.performanceanalyzer.config.PerformanceAnalyzerController;
+import com.amazon.opendistro.elasticsearch.performanceanalyzer.util.Utils;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.elasticsearch.action.admin.indices.stats.CommonStatsFlags;
+import org.elasticsearch.action.admin.indices.stats.IndexShardStats;
+import org.elasticsearch.action.admin.indices.stats.ShardStats;
+import org.elasticsearch.index.shard.IndexShard;
+import org.elasticsearch.index.shard.ShardId;
+import org.elasticsearch.indices.IndicesService;
+import org.elasticsearch.indices.NodeIndicesStats;
+import com.amazon.opendistro.elasticsearch.performanceanalyzer.ESResources;
+import com.amazon.opendistro.elasticsearch.performanceanalyzer.metrics.AllMetrics.ShardStatsValue;
+import com.amazon.opendistro.elasticsearch.performanceanalyzer.metrics.MetricsConfiguration;
+import com.amazon.opendistro.elasticsearch.performanceanalyzer.metrics.MetricsProcessor;
+import com.amazon.opendistro.elasticsearch.performanceanalyzer.metrics.PerformanceAnalyzerMetrics;
+import com.fasterxml.jackson.annotation.JsonIgnore;
+import com.fasterxml.jackson.annotation.JsonProperty;
+
+/**
+ * This collector collects metrics for all shards on a node in a single run.
+ * These metrics are light weight metrics which have minimal performance impacts
+ * on the performance of the node.
+ */
+
+@SuppressWarnings("unchecked")
+public class NodeStatsAllShardsMetricsCollector extends PerformanceAnalyzerMetricsCollector implements MetricsProcessor {
+    public static final int SAMPLING_TIME_INTERVAL = MetricsConfiguration.CONFIG_MAP.get(
+            NodeStatsAllShardsMetricsCollector.class).samplingInterval;
+    private static final int KEYS_PATH_LENGTH = 2;
+    private static final Logger LOG = LogManager.getLogger(NodeStatsAllShardsMetricsCollector.class);
+    private HashMap<String, IndexShard> currentShards;
+    private final PerformanceAnalyzerController controller;
+
+
+    public NodeStatsAllShardsMetricsCollector(final PerformanceAnalyzerController controller) {
+        super(SAMPLING_TIME_INTERVAL, "NodeStatsMetrics");
+        currentShards = new HashMap<>();
+        this.controller = controller;
+    }
+
+    private String getUniqueShardIdKey(ShardId shardId) {
+        return "[" + shardId.getIndex().getUUID() + "][" + shardId.getId() + "]";
```

Comment:
I made a mistake when writing this function. Let's take this opportunity to correct it and use the shardId.hashCode() instead.

Reply:
Done